### DISCUSSION TITLE
Querying for data

### DISCUSSION THREAD

**Discussion by joeldrapper** [yippee-fun/phlex#29] [Ideas]
I think we have an opportunity to design a pattern that significantly improves safety and performance when querying for data to render into views. Here's a profile of a typical Rails view:

![FXYBJVSWYAEH9Io](https://user-images.githubusercontent.com/246692/179270991-0cf4aa4c-cd41-4e5b-baca-39fa230d3b76.jpeg)

Rails starts rendering the views but has to stop several times to query the database. Some of these queries haven't even been issued to the database until about half way through the response.

Ideally, each of these database queries would have been issued as early as possible — either as the first thing that happened when the controller action was hit or immediately after a contingent query had finished.

I think we can achieve this by wrapping queries in [Ractors](https://rubyapi.org/3.1/o/ractor) which are then consumed by subsequent queries and/or components. The components can call `take` on the query ractors when they need the data. We should also render components in Ractors so they don't themselves need to wait on sibling components.

Additionally, data should be immutable from within a view. Views can render forms, but they should never directly mutate data.

**Comment by fractaledmind** [comment]
Is my understanding of StaticComponents correct, such that if a Component is a StaticComponent, it cannot contain queries and cannot mutate data?

**Reply by joeldrapper** [reply]
No, static components can include dynamic data, though as things stand in #20, they cannot contain logic.

**Comment by joeldrapper** [comment]
I think the new `load_async` feature in ActiveRecord pretty much solves this. It'd be nice if they could be frozen too.
